snap/epoch.go

```diff
@@ -0,0 +1,288 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snap
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+)
+
+var (
+	ErrBadEpochExpression  = errors.New("invalid epoch expression")
+	ErrEpochOhSplat        = errors.New("0* is an invalid epoch")
+	ErrBadEpochNumber      = errors.New("epoch numbers must match [1-9][0-9]*")
+	ErrHugeEpochNumber     = errors.New("epoch numbers must be less than 2³²")
+	ErrBadEpochList        = errors.New("epoch read/write attributes must be lists of epoch numbers")
+	ErrEmptyEpochList      = errors.New("epoch list cannot be explicitly empty")
+	ErrEpochListNotSorted  = errors.New("epoch list must be in ascending order")
+	ErrNoEpochIntersection = errors.New("epoch read and write lists must have a non-empty intersection")
+)
+
+// An Epoch represents the ability of the snap to read and write its data. Most
+// developers need not worry about it, and snaps default to the 0th epoch, and
+// users are only offered refreshes to epoch 0 snaps. Once an epoch bump is in
+// order, there's a simplified expression they can use which should cover the
+// majority of the cases:
+//
+//   epoch: N
+//
+// means a snap can read/write exactly the Nth epoch's data, and
+//
+//   epoch: N*
+//
+// means a snap can additionally read (N-1)th epoch's data, which means it's a
+// snap that can migrate epochs (so a user on epoch 0 can get offered a refresh
+// to a snap on epoch 1*).
+//
+// If the above is not enough, a developer can explicitly describe what epochs a
+// snap can read and write:
+//
+//   epoch:
+//     read: [1, 2, 3]
+//     write: [1, 3]
+//
+// the read attribute defaults to the value of the write attribute, and the
+// write attribute defaults to the last item in the read attribute. If both are
+// unset, it's the same as not specifying an epoch at all (i.e. epoch: 0). The
+// lists must be in ascending order, and there must be a non-empty intersection
+// between them.
+//
+// Epoch numbers must be written as base 10 numbers, with no zero padding.
+type Epoch struct {
+	Read  []uint32 `yaml:"read"`
+	Write []uint32 `yaml:"write"`
+}
+
+// E returns the epoch represented by the expression s. It's meant for use in
+// testing, as it panics at the first sign of trouble.
+func E(s string) Epoch {
```

Comment:
Using Epoch by value is not so useful and might easily be misleading. The typical reason to have that is for example time.Time, which you can hand around and preserve the value semantics, or in our case of snap.Revision, which follows that same logic. In the case of epochs, though, its whole content is mutable, so it is handled by identity even if we're copying it by value. Easy foot-shooting. I suggest tuning all the logic bellow (returns and method receivers) so that Epoch is always a pointer. It's fine that it remains a value in the snap struct, though, because the struct itself is handle as a pointer.

Reply:
done